subscription email: do not send email if nothing has changed

saschaszott opened the pull request:
## Description

This PR makes the email subscription mechanism more robust. It adds a check whether the subscription email contains at least one item. In case it does not find any new items the subscription email is not sent to the subscriber.

Furthermore, this PR improves the Velocity template by adding checks for empty strings to improve readability of the generated subscription email.

saschaszott committed:
improved subscriptions email template

saschaszott committed:
do not send emails without content

tdonohue commented:
@saschaszott : This PR is failing unit tests because of checkstyle violations.  If you run `mvn checkstyle:check` locally you should be able to see them.  Here's what is reported:
```
[INFO] There are 2 errors reported by Checkstyle 8.30 with /home/runner/work/DSpace/DSpace/dspace-api/../checkstyle.xml ruleset.
Error:  src/main/java/org/dspace/subscriptions/ContentGenerator.java:[63] (sizes) LineLength: Line is longer than 120 characters (found 143).
Error:  src/main/java/org/dspace/subscriptions/ContentGenerator.java:[68] (regexp) RegexpSingleline: Line has trailing whitespace
```

Both should be easy to cleanup

saschaszott committed:
fixed coding style violations

saschaszott commented:
> @saschaszott : This PR is failing unit tests because of checkstyle violations. If you run `mvn checkstyle:check` locally you should be able to see them.
> 
> Both should be easy to cleanup

@tdonohue , thanks for the hint. I have fixed the coding style violations and added Checkstyle to my development setup.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/subscriptions/ContentGenerator.java:
```diff
@@ -70,7 +78,7 @@ private String generateBodyMail(Context context, List<IndexableObject> indexable
             ByteArrayOutputStream out = new ByteArrayOutputStream();
             out.write("\n".getBytes(UTF_8));
-            if (indexableObjects.size() > 0) {
+            if (indexableObjects != null && !indexableObjects.isEmpty()) {
```
`.isEmpty()` check should be unnecessary.  If the List is empty, the `for` loop should exit immediately.

Perhaps the `null` test should be the first thing done, returning `""` if true.

tdonohue commented:
@saschaszott : If you find time in the next week or so, it'd be good to get this PR updated based on @mwoodiupui 's [feedback above](https://github.com/DSpace/DSpace/pull/8981#pullrequestreview-1598759539).  We are working on getting 7.6.1 finished up in October (if possible), in order to release in early to mid Nov.  I'd love to include this PR if it can be cleaned up in time.  Thanks!

alanorth commented:
Thanks @saschaszott. Is there an easy way to test this? We have thousands of users and SMTP support is disabled on my development server for obvious reasons. When I run the `dspace subscriptions -f D` command for daily subscriptions I just get tens of thousands of lines about SMTP errors in my `dspace.log`.

tdonohue commented:
@alanorth : In development, if you have `mail.server.disabled = true` set, then the email text will be logged in your `dspace.log` file.  That might be used to verify if an email _would have been sent_ or not.  If it's in the logs, it would have gone out.  If not, then it would not go out.

saschaszott committed:
removed unnecessary isEmpty check as suggested by reviewer

saschaszott committed:
moved null check on indexableObjects in generateBodyMail

saschaszott commented:
@tdonohue , @mwoodiupui I've fixed the proposed changes in `ContentGenerator`.

saschaszott committed:
fixed unhandled IOException

tdonohue commented on dspace-api/src/main/java/org/dspace/subscriptions/ContentGenerator.java:
```diff
@@ -56,8 +56,16 @@ public void notifyForSubscriptions(Context context, EPerson ePerson,
+                    return;
+                }
+                email.addArgument(bodyCommunitites);
+                email.addArgument(bodyCollections);
```
This code appears to be run even if one of `bodyCommunities` or `bodyCollections` is empty... so, it is still possible to get an empty section, as long as one section has content.

tdonohue commented on dspace-api/src/main/java/org/dspace/subscriptions/ContentGenerator.java:
```diff
@@ -56,8 +56,16 @@ public void notifyForSubscriptions(Context context, EPerson ePerson,
-                email.addArgument(generateBodyMail(context, indexableComm));
-                email.addArgument(generateBodyMail(context, indexableColl));
+
+                String bodyCommunitites = generateBodyMail(context, indexableComm);
```
Tiny thing. Variable is misspelled.  Should be `bodyCommunities`

tdonohue reviewed (commented):
@saschaszott : I tested this today, and it seems to work overall.  Though, I'm not sure the changes to the `subscriptsion_content` template are working, as I still find I'm getting empty sections.

For example, my setup is as follows:
* Add a new "Daily" Subscription to a Community (only) which has Items added in the last day.
* Run `subscription-send -f D`.
* In the email, I see the items under the Community section, but also have an *empty* Collection section (even though I have no Collection subscriptions).  It looks like this:
```
Date: Thu, 26 Oct 2023 18:36:21 +0000 (UTC)
From: dspace-noreply@myu.edu
To: dspacedemo+admin@gmail.com
Subject: DSpace Started with Docker Compose Subscriptions

This email is sent from DSpace Started with Docker Compose based on the chosen subscription preferences.

Community Subscriptions:
------------------------
List of changed items :

DSpace logo uploaded via e2e tests http://localhost:4000/handle/123456789/1191

Test PowerPoint Document http://localhost:4000/handle/123456789/1122

Oliver Twist http://localhost:4000/handle/123456789/1175

Collection Subscriptions:
 -------------------------
List of changed items :

```

NOTICE the empty "Collection Subscriptions" section above. I have verified that I have no subscriptions to any collection, so it's odd that I see that section.  
I suspect the reason is that the logic in ContentGenerator seems to be adding *both* sections as long as *one* of them is non-empty.  I've noted this logic below.

Beyond that, the behavior seems to be *improved*.  It's just not doing exactly what it claims to do, as it is still possible to have empty sections (so the changes to `subscriptions_content` seem to have no effect?)

saschaszott committed:
fixed typo in bodyCommunities

saschaszott committed:
do not use != to compare strings

tdonohue commented:
@saschaszott : I retested this today since I saw you made updates.  I'm still unfortunately seeing the same behavior I described above.  If I subscribe to *only* one Community, I will still see an **empty** "Collection Subscriptions" section in every email report.

So, this doesn't appear to be working as described.  My expectation is that empty sections should no longer appear, but I'm able to get an empty "Collection Subscriptions" section in *every email* just by doing this:
1. Subscribe to a single Community with a "Daily" subscription.  Make sure this is your own subscription
2. Submit a new item into a Collection within that Community
3. Run `subscription-send -f D`.  In the email, you should see the Item you created in step 2 under "Community Subscriptions". But you will also see an **empty** "Collection Subscriptions" section.

These are the same steps[ I described above](https://github.com/DSpace/DSpace/pull/8981#pullrequestreview-1700382080) as well.

saschaszott commented:
@tdonohue, this is strange. I have no idea. Can you explain the additional leading space in the email content you've pasted in the comment https://github.com/DSpace/DSpace/pull/8981#pullrequestreview-1700382080

```
Collection Subscriptions:
 -------------------------
List of changed items :
```

Did you consider my recent change (https://github.com/DSpace/DSpace/pull/8981/commits/e46018333508215fce7225b75a5af50d4d7beb59) to the Velocity template (check for empty string)?

saschaszott commented:
Ok, sounds reasonable - I have to admit that I've tested the proposed bugfix in our DS CRIS 2023.01 installation and it works. Maybe there is some subtle difference in the "surrounding" subscription implementation (e.g., in class `org.dspace.core.Email`)?

saschaszott committed:
fixed improper handling of empty list

tdonohue reviewed (approved):
👍 Thanks @saschaszott !  That small fix resolved the issues for me!  This now works properly and the code looks good.

tdonohue commented:
While this is a very late change, merging immediately to include in 7.6.1.  This is a small PR overall and very specific to subscription emails.
